src/main/java/com/artipie/asto/Remaining.java

```diff
@@ -54,9 +69,25 @@ public Remaining(final ByteBuffer buf) {
      */
     public byte[] bytes() {
         final byte[] bytes = new byte[this.buf.remaining()];
-        this.buf.mark();
+        if (this.restore) {
+            this.buf.mark();
+        }
         this.buf.get(bytes);
-        this.buf.reset();
+        if (this.restore) {
+            this.buf.reset();
+        }
         return bytes;
     }
+
+    /**
+     * Buffer read strategy.
+     * @since 0.14
+     */
+    public interface ReadStrategy {
+        /**
+         * Apply strategy to data buffer.
+         * @param data Output data
+         */
+        void apply(byte[] data);
```

Comment:
@g4s8 I don't  understand why passing a `byte[]` here. IMHO this strategy would be responsible for handling the type of read on the buffer (with restore / without restore), so why don't pass the buffer itself to this method?